substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,763 @@
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Existing `agent` cannot register again and a delegator cannot become an `agent`.
+			ensure!(!Self::is_agent(&who) && !Self::is_delegator(&who), Error::<T>::NotAllowed);
+
+			// They cannot be already a direct staker in the staking pallet.
+			ensure!(!Self::is_direct_staker(&who), Error::<T>::AlreadyStaking);
+
+			// Reward account cannot be same as `agent` account.
+			ensure!(reward_account != who, Error::<T>::InvalidRewardDestination);
+
+			Self::do_register_agent(&who, &reward_account);
+			Ok(())
+		}
+
+		/// Migrate from a `Nominator` account to `Agent` account.
+		///
+		/// The origin needs to
+		/// - be a `Nominator` with [`Config::CoreStaking`],
+		/// - not already an `Agent`,
+		/// - have enough funds to transfer existential deposit to a delegator account created for
+		///   the migration.
+		///
+		/// This function will create a proxy account to the agent called `proxy_delegator` and
+		/// transfer the directly staked amount by the agent to it. The `proxy_delegator` delegates
+		/// the funds to the origin making origin an `Agent` account. The real `delegator`
+		/// accounts of the origin can later migrate their funds using [Self::claim_delegation] to
+		/// claim back their share of delegated funds from `proxy_delegator` to self.
+		pub fn migrate_to_agent(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// ensure who is a staker in `CpreStaking` but not already an agent.
+			ensure!(!Self::is_agent(&who) && Self::is_direct_staker(&who), Error::<T>::NotAllowed);
+
+			// Reward account cannot be same as `agent` account.
+			ensure!(reward_account != who, Error::<T>::InvalidRewardDestination);
+
+			Self::do_migrate_to_agent(&who, &reward_account)
+		}
+
+		/// Release previously delegated funds by delegator to origin.
+		///
+		/// Only agents can call this.
+		///
+		/// Tries to withdraw unbonded funds from `CoreStaking` if needed and release amount to
+		/// `delegator`.
+		pub fn release_delegation(
+			origin: OriginFor<T>,
+			delegator: T::AccountId,
+			amount: BalanceOf<T>,
+			num_slashing_spans: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Self::do_release(&who, &delegator, amount, num_slashing_spans)
+		}
+
+		/// Claim delegated funds that are held in `proxy_delegator` to the claiming delegator's
+		/// account. If successful, the specified funds will be delegated directly from `delegator`
+		/// account to the agent.
+		///
+		/// This can be called by `agent` accounts that were previously a direct `Nominator` with
+		/// [`Config::CoreStaking`] and has some remaining unclaimed delegations.
+		///
+		/// Internally, it moves some delegations from `pxoxy_delegator` account to `delegator`
+		/// account and reapplying the holds.
+		pub fn claim_delegation(
```

Comment:
How about renaming this to `fn migrate_delegation`? Or else rename `fn do_migrate_delegation` to `fn do_claim_delegation`.

Reply:
No strong opinions though the way I was thinking, the public function naming indicates what this function does from delegator's point of view. Internally it is a migration of delegated amount.